Validate parse trees in dialogue manager + modules path fix

Rebecca-Qian opened the pull request:
# Description

Add a hook in dialogue manager to validate parse trees from ground truth or semantic parser.

If the parse tree fails the grammar spec, treat as NOOP (log details).

Added a new dialogue manager unit test to test `validate_parse_tree` method. 

PR also contains some fixes for module paths, see #117. 

Fixes # (issue)
#145 
#117 (partial fix)

## Type of change

Please check the options that are relevant.

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [x] New feature (non-breaking change that adds a new functionality)
- [ ] Breaking change (fix or feature that would break some existing functionality downstream)
- [x] This is a unit test
- [ ] Documentation only change

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

Example of action dictionary caught by validator when playing with bot (later fixed in ground truth):
P176155293

# Testing

In addition to conda setup, make sure modules are properly initialized with
`python setup.py develop`

Then run unit tests and test the bot as usual.

# Checklist:

- [x] I have performed manual end-to-end testing of the feature in my environment.
- [x] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [x] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [x] I have added relevant collaborators to review the PR before merge.

Rebecca-Qian commented:
FYI, removed `spacy` from `craftassist/requirements.txt` because it conflicts with base_agent dependencies. See P176017299 for error trace.

Doesn't seem to cause issues for me, cc @snyxan @kavyasrinet

Rebecca-Qian commented:
Update: also fixed import paths in locobot tests, since some tests were failing.

cc @anuragprat1k

anuragprat1k commented on base_agent/nsp_dialogue_manager.py:
```diff
@@ -269,6 +294,16 @@ def get_logical_form(self, s: str, chat_as_list=False, ground_truth_actions: dic
                 d = self.model.parse(chat=s)
 
+        # Validate parse tree against grammar
+        is_valid_json = self.validate_parse_tree(d)
```
how frequently do you expect this to be invalid? do we want to collect all the invalid parses - if so how?

Rebecca-Qian replied:
For ground truth, this should never be invalid, if it is, we need to fix ASAP. For parses from the semantic parser, it should not happen frequently. The model is generally good at producing parse trees that fit the grammar spec, even when they're wrong.

Agree, the wrong parses should be collected. I'll write to a separate file in datasets.

Rebecca-Qian replied:
Actually, may not need to write the failed parses to a separate file. Considering adding data checks to CI, then ground truth would always be up to date. Failed model parses aren't very actionable

anuragprat1k commented on craftassist/test/test_dialogue_manager.py:
```diff
@@ -80,6 +80,11 @@ def test_parses(self):
                 )
             )
 
+    def test_validate_json(self):
```
add a successful validation test scenario as well.

anuragprat1k commented on craftassist/test/validate_json.py:
```diff
@@ -27,7 +27,16 @@ def validate_data(data_path, schema, resolver):
+                print("\n")
+
+
+def validate_instance(parse_tree, schema, resolver):
```
where is this used?

Rebecca-Qian replied:
Added docstring here

anuragprat1k reviewed (changes requested):
The json validation changes look good. Requested some minor changes.

I don't see any of the `sys.path.append`s removed in this PR, so I don't understand why the tests were failing - did you figure out why?

kavyasrinet reviewed (commented):
Thanks for making the change. Looks good. Let's add a unit test too.

Rebecca-Qian commented:
Updates:

- Added 2 test cases checking for spans with string values and arrays
- Added support for string value spans, since these are the format of spans in unit tests and currently autocomplete outputs. I think it's more readable to keep hand written parses in this format, so I decided to support different validation formats rather than enforce array spans everywhere.
- Refactored JSON validator and added docstrings

anuragprat1k reviewed (approved):
looks good.
